networkx/algorithms/approximation/maxcut.py

```diff
@@ -0,0 +1,117 @@
+
+
+@nx.not_implemented_for("directed", "multigraph")
+@py_random_state(1)
+def randomized_partitioning(G, seed=None, p=0.5, weight=None):
+    """Compute a random partitioning of the graph nodes and its cut value.
+
+    A partitioning is calculated by observing each node
+    and deciding to add it to the partition with probability p,
+    returning a random cut and its corresponding value (the
+    sum of weights of edges connecting different partitions).
+
+    Parameters
+    ----------
+    G: NetworkX graph
+
+    seed : integer, random_state, or None (default)
+        Indicator of random number generation state.
+        See :ref:`Randomness<randomness>`.
+
+    p: double@nx.not_implemented_for('directed', 'multigraph')
+        Probability for each node to be part of the first partition.
+        Should be in [0,1]
+
+    weight : object
+        Edge attribute key to use as weight. If not specified, edges
+        have weight one.
+
+    Returns
+    -------
+    cut_size : integer, float
+        Value of the minimum cut.
+
+    partition : pair of node sets
+        A partitioning of the nodes that defines a minimum cut.
+    """
+    cut = set()
+    for node in G.nodes():
+        if seed.random() < p:
+            cut.add(node)
+    cut_size = nx.algorithms.cut_size(G, cut, weight=weight)
+    partition = (cut, G.nodes - cut)
+    return cut_size, partition
+
+
+def _swap_node_partition(cut, node):
+    if node in cut:
+        new_cut = cut - {node}
+    else:
+        new_cut = cut.union({node})
+    return new_cut
+
+
+@nx.not_implemented_for("directed", "multigraph")
+@py_random_state(2)
+def one_exchange(G, initial_cut=None, seed=None, weight=None):
+    """Compute a partitioning of the graphs nodes and the corresponding cut value.
+
+    Use a greedy one exchange strategy to find a locally maximal cut
+    and its value, it works by finding the best node (one that gives
+    the highest gain to the cut value) to add to the current cut
+    and repeats this process until no improvement can be made.
+
+    Parameters
+    ----------
+    G: networkx Graph
+        Graph to find a maximum cut for.
+
+    initial_cut: set
+        Cut to use as a starting point. If not supplied the algorithm
+        starts with an empty cut.
+
+    seed : integer, random_state, or None (default)
+        Indicator of random number generation state.
+        See :ref:`Randomness<randomness>`.
+
+    weight : object
+        Edge attribute key to use as weight. If not specified, edges
+        have weight one.
+
+    Returns
+    -------
+    cut_value : integer, float
+        Value of the maximum cut.
+
+    partition : pair of node sets
+        A partitioning of the nodes that defines a maximum cut.
+    """
+    if initial_cut is None:
+        initial_cut = set()
+    cut = set(initial_cut)
+    current_cut_size = nx.algorithms.cut_size(G, cut, weight=weight)
+    while True:
+        nodes = list(G.nodes())
+        seed.shuffle(nodes)
```

Comment:
It looks like `nodes` in only used in the subsequent `max` operation, so does the ordering of the nodes matter?

Reply:
It matters because of possible ties in the max call. As max always returns the first element with maximum key[1], the shuffling can yield to completely different solutions.

[1]: https://docs.python.org/3/library/functions.html#max

Reply:
Gotcha - thanks for the explanation. Out of curiosity, is the answer wrong without the shuffle, or just different? I'm not familiar with the algorithm so this is mostly just to satisfy my own curiosity :)

Since it's not immediately obvious from the code, it might be worth a comment above the `shuffle` line explaining its role in the algorithm. Not a blocker though - whatever you think!

Reply:
Thanks for your feedback. The algorithm still returns correct results without the shuffle, that is a valid node bipartition. But in a way the shuffling increases the influence of the seed and allows the algorithm to find different (possibly better) solutions for the max-cut problem. This is helpful when calling the algorithm multiple times with different seeds in search for a good/better solution. I added a comment providing basic reasoning for the shuffle.

Reply:
Makes a lot of sense, thanks!